extended_tasks/mt_bench/judges.py

```diff
@@ -0,0 +1,136 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+import ast
+import json
+import re
+import time
+from abc import ABC
+from typing import Optional
+
+from openai import OpenAI
+
+from lighteval.logging.hierarchical_logger import hlog_warn
+
+
+# Abstract class for a judge
+class Judge(ABC):
+    def evaluate_answer(answers, questions, references) -> tuple[str, list[dict[str, str]], str]:
+        pass
+
+
+class JudgeOpenAI(Judge):
+    def __init__(self, model: str, seed: int, temperature: float, templates_path: str):
+        self.client = OpenAI()
+        self.model = model
+        self.seed = seed
+        self.temperature = temperature
+
+        data = []
+        with open(templates_path, "r") as f:
+            for line in f:
+                tmp = json.loads(line)
+                data.append(tmp)
+
+        self.templates = {d["name"]: d for d in data}
+
+        self.one_score_pattern = re.compile(r"\[\[(\d+\.?\d*)\]\]")
+        self.one_score_pattern_backup = re.compile(r"\[(\d+\.?\d*)\]")
+
+        self.API_MAX_RETRY = 16
+        self.API_RETRY_SLEEP = 10
+        self.max_tokens = 2048
+
+    def evaluate_answer(
+        self, questions: list[str], answers: list[str], references: list[str], single_turn: bool
+    ) -> tuple[int, list[dict[str, str]], str]:
+        if single_turn:
+            prompts = self.__get_prompts_single_turn(
+                questions[0], answers[0], references[0] if len(references) > 0 else None
+            )
+        else:
+            prompts = self.__get_prompts_multi_turn(questions, answers, references if len(references) > 0 else None)
+
+        for _ in range(self.API_MAX_RETRY):
+            try:
+                response = self.client.chat.completions.create(
+                    model=self.model,
+                    seed=self.seed,
+                    temperature=self.temperature,
+                    messages=prompts,
+                    max_tokens=self.max_tokens,
+                    n=1,
+                )
+                break
+            except Exception as e:
+                hlog_warn(f"{type(e), e}")
+                time.sleep(self.API_RETRY_SLEEP)
+
+        judgment = response.choices[0].message.content
```

Comment:
What happens here if it failed all the times?

Reply:
Maybe print an error message and stop the script

Reply:
Yeah, it's done just above, I raise an exception after having printed the reason for the fail.